src/Conversion/ONNXToKrnl/ConvertONNXToKrnl.cpp

```diff
@@ -33,17 +33,138 @@ class ONNXEntryPointLowering : public OpRewritePattern<ONNXEntryPointOp> {
+    func::FuncOp entryPointFunc = cast<func::FuncOp>(entryPointOp);
+
+    IntegerAttr numInputsAttr =
+        rewriter.getI32IntegerAttr(entryPointFunc.getArgumentTypes().size());
+    IntegerAttr numOutputsAttr =
+        rewriter.getI32IntegerAttr(entryPointFunc.getResultTypes().size());
+    std::string sig =
+        getSignature(entryPointFunc.getFunctionType(), entryPointOp);
+    StringAttr sigAtrr = rewriter.getStringAttr(sig);
+
+    rewriter.replaceOpWithNewOp<KrnlEntryPointOp>(
+        op, funcRefAttr, numInputsAttr, numOutputsAttr, sigAtrr);
     return success();
   }
+
+private:
+  // construct JSON type from the argument type
+  // for example - a 3D array of f32 would produce something like
+  //     {"type" : "f32" , "dims" : [4, 256, 16] , "name": "t1"}
+  // data type list:
+  //     "i1" / "i8" / "i16" / "i32" / "i64"
+  //     "ui8" / "ui16" / "ui32" / "ui64"
+  //     "f32" / "f64"
+  void concatTypeString(
+      Type argType, Attribute attr, llvm::raw_ostream &dstream) const {
+    std::string comma = std::string("");
+
+    TypeSwitch<Type>(argType)
+        .Case<mlir::SeqType>([&](mlir::SeqType seqTy) {
+          auto et = seqTy.getElementType();
+          dstream << "   {\"seq\" : ";
+          concatTypeString(et, attr, dstream);
+        })
+        .Case<ShapedType>([&](ShapedType tensorTy) {
+          auto et = tensorTy.getElementType();
+          dstream << "   { \"type\" : ";
+          et.print(dstream);
+          dstream << " , \"dims\" : [";
+          if (tensorTy.hasRank()) {
+            int64_t rank = tensorTy.getRank();
+            for (int j = 0; j < rank; j++) {
+              dstream << comma << tensorTy.getDimSize(j);
+              comma = std::string(" , ");
+            }
+          } else {
+          }
+          dstream << "] ";
+          auto name = attr.cast<mlir::StringAttr>().getValue().str();
+          dstream << ", \"name\" : \"" << name << "\"";
+        })
+        .Default([&](Type type) { llvm_unreachable("input is not a tensor"); });
+    dstream << " }\n";
+  }
+
+  std::string getSignature(FunctionType funcType, Operation *op) const {
+    OpBuilder b(op);
+    auto inputs = funcType.getInputs();
+    auto outputs = funcType.getResults();
+
+    ArrayAttr inputNames = op->getAttrOfType<ArrayAttr>("input_names");
+    if (!inputNames) {
+      SmallVector<StringRef, 4> names;
+      for (uint64_t i = 0; i < inputs.size(); ++i)
+        names.emplace_back(StringRef("input_" + std::to_string(i)));
+      inputNames = b.getStrArrayAttr(names);
+    }
+    ArrayAttr outputNames = op->getAttrOfType<ArrayAttr>("output_names");
+    if (!outputNames) {
+      SmallVector<StringRef, 4> names;
+      for (uint64_t i = 0; i < outputs.size(); ++i)
+        names.emplace_back(StringRef("output_" + std::to_string(i)));
+      outputNames = b.getStrArrayAttr(names);
+    }
+
+    std::string const sf32 = std::string(" f32 ");
```

Comment:
suggestion: each of the sf64... seemed to be used only once. Could you define it directly in the map? Also, you could make the map static, and possibly have a `static bool isMapInit=false` and initialize on first call if its not possible to have the init of the map directly in a constructor.

Reply:
Thanks! I puts the strings directly into the map and make the map static.